File: Sources/PDFLocalCertKit/CoordinateMapper.swift
import CoreGraphics

/// Geometry bridge between the three coordinate spaces involved in placing a
/// visible signature:
///
///   1. **View space** — SwiftUI/AppKit points over the rendered page, origin
///      top-left, +y downward. What the user actually drags.
///   2. **Normalized displayed space** — fractions 0…1 of the *displayed* page
///      box (after `/Rotate` is applied by the viewer), origin top-left. Stored
///      as a resolution/zoom-independent description of the box.
///   3. **PDF user space** — the unrotated coordinate system `/Rect` lives in:
///      origin bottom-left, +y upward, measured in points, offset by the
///      cropBox origin. This is what the Rust core writes into the annotation.
///
/// The mapper is intentionally free of PDFKit/UI types so the rotation × zoom
/// matrix can be unit-tested with plain values. A `PDFPage` convenience init is
/// provided behind `canImport(PDFKit)` for the app target.
public struct CoordinateMapper: Sendable, Equatable {
    /// The page's cropBox in PDF user space (origin may be non-zero).
    public let cropBox: CGRect
    /// `/Rotate` value, normalized to one of 0, 90, 180, 270.
    public let rotation: Int

    public init(cropBox: CGRect, rotation: Int) {
        self.cropBox = cropBox
        self.rotation = ((rotation % 360) + 360) % 360
    }

    /// Displayed (post-rotation) page size in points.
    public var displayedSize: CGSize {
        switch rotation {
        case 90, 270: return CGSize(width: cropBox.height, height: cropBox.width)
        default: return CGSize(width: cropBox.width, height: cropBox.height)
        }
    }

    // MARK: View ↔ normalized

    /// Convert a rect in view space (origin top-left) into a normalized rect,
    /// given the frame the page occupies in that same view space.
    /// The result is clamped to 0…1 on each axis.
    public func normalize(viewRect: CGRect, in pageFrame: CGRect) -> CGRect {
        guard pageFrame.width > 0, pageFrame.height > 0 else { return .zero }
        let nx = (viewRect.minX - pageFrame.minX) / pageFrame.width
        let ny = (viewRect.minY - pageFrame.minY) / pageFrame.height
        let nw = viewRect.width / pageFrame.width
        let nh = viewRect.height / pageFrame.height
        return clamp01(CGRect(x: nx, y: ny, width: nw, height: nh))
    }

    /// Inverse of `normalize` — place a normalized rect back into a view frame.
    public func viewRect(normalized n: CGRect, in pageFrame: CGRect) -> CGRect {
        CGRect(
            x: pageFrame.minX + n.minX * pageFrame.width,
            y: pageFrame.minY + n.minY * pageFrame.height,
            width: n.width * pageFrame.width,
            height: n.height * pageFrame.height)
    }

    // MARK: Normalized ↔ PDF user space

    /// Map a normalized displayed rect (origin top-left) to a PDF user-space
    /// rect (origin bottom-left, unrotated, cropBox-offset) for `/Rect`.
    public func userSpaceRect(normalized n: CGRect) -> CGRect {
        let d = displayedSize
        // Normalized (top-left) → displayed points (bottom-left origin).
        let dx = n.minX * d.width
        let dw = n.width * d.width
        let dh = n.height * d.height
        let dy = d.height - (n.minY * d.height) - dh // flip y
        // The two opposite corners of the box in displayed space.
        let c0 = displayedToLocal(CGPoint(x: dx, y: dy))
        let c1 = displayedToLocal(CGPoint(x: dx + dw, y: dy + dh))
        let minX = min(c0.x, c1.x), minY = min(c0.y, c1.y)
        let maxX = max(c0.x, c1.x), maxY = max(c0.y, c1.y)
        return CGRect(
            x: cropBox.minX + minX,
            y: cropBox.minY + minY,
            width: maxX - minX,
            height: maxY - minY)
    }

    /// Inverse: PDF user-space rect → normalized displayed rect (top-left).
    public func normalizedRect(userSpace r: CGRect) -> CGRect {
        let local0 = CGPoint(x: r.minX - cropBox.minX, y: r.minY - cropBox.minY)
        let local1 = CGPoint(x: r.maxX - cropBox.minX, y: r.maxY - cropBox.minY)
        let d0 = localToDisplayed(local0)
        let d1 = localToDisplayed(local1)
        let dminX = min(d0.x, d1.x), dmaxX = max(d0.x, d1.x)
        let dminY = min(d0.y, d1.y), dmaxY = max(d0.y, d1.y)
        let d = displayedSize
        guard d.width > 0, d.height > 0 else { return .zero }
        let nx = dminX / d.width
        let nw = (dmaxX - dminX) / d.width
        let nh = (dmaxY - dminY) / d.height
        // Flip y back to top-left origin.
        let ny = 1.0 - (dmaxY / d.height)
        return CGRect(x: nx, y: ny, width: nw, height: nh)
    }

    // MARK: Rotation core (page-local, bottom-left origin, both unrotated px,py)

    /// Displayed-space point (bottom-left origin) → page-local point
    /// (unrotated cropBox-local, bottom-left origin).
    private func displayedToLocal(_ p: CGPoint) -> CGPoint {
        let w = cropBox.width, h = cropBox.height
        switch rotation {
        case 90:  return CGPoint(x: w - p.y, y: p.x)
        case 180: return CGPoint(x: w - p.x, y: h - p.y)
        case 270: return CGPoint(x: p.y, y: h - p.x)
        default:  return p
        }
    }

    /// Page-local point → displayed-space point (inverse of `displayedToLocal`).
    private func localToDisplayed(_ p: CGPoint) -> CGPoint {
        let w = cropBox.width, h = cropBox.height
        switch rotation {
        case 90:  return CGPoint(x: p.y, y: w - p.x)
        case 180: return CGPoint(x: w - p.x, y: h - p.y)
        case 270: return CGPoint(x: h - p.y, y: p.x)
        default:  return p
        }
    }

    private func clamp01(_ r: CGRect) -> CGRect {
        let x = max(0, min(1, r.minX))
        let y = max(0, min(1, r.minY))
        let w = max(0, min(1 - x, r.width))
        let h = max(0, min(1 - y, r.height))
        return CGRect(x: x, y: y, width: w, height: h)
    }
}

File: windows/BureaucratPdf.Core/CoordinateMapper.cs
namespace BureaucratPdf.Core;

/// <summary>
/// Geometry bridge between the three coordinate spaces involved in placing a
/// visible signature. Ported verbatim from BureaucratPdfKit.CoordinateMapper
/// (apple/Sources/BureaucratPdfKit/CoordinateMapper.swift) — the math must stay
/// byte-identical so both shells place signatures in the same spot (anti-desync).
///
///   1. View space          — points over the rendered page, origin top-left, +y down.
///   2. Normalized displayed — fractions 0..1 of the *displayed* page box (after
///                             /Rotate), origin top-left. Zoom/resolution independent.
///   3. PDF user space       — the unrotated system /Rect lives in: origin bottom-left,
///                             +y up, points, offset by the cropBox origin.
/// </summary>
public readonly struct CoordinateMapper : IEquatable<CoordinateMapper>
{
    /// <summary>The page's cropBox in PDF user space (origin may be non-zero).</summary>
    public PdfRect CropBox { get; }

    /// <summary>/Rotate value, normalized to one of 0, 90, 180, 270.</summary>
    public int Rotation { get; }

    public CoordinateMapper(PdfRect cropBox, int rotation)
    {
        CropBox = cropBox;
        Rotation = ((rotation % 360) + 360) % 360;
    }

    /// <summary>Displayed (post-rotation) page size in points.</summary>
    public PdfSize DisplayedSize => Rotation switch
    {
        90 or 270 => new PdfSize(CropBox.Height, CropBox.Width),
        _ => new PdfSize(CropBox.Width, CropBox.Height),
    };

    // ── View ↔ normalized ────────────────────────────────────────────────────

    /// <summary>
    /// Convert a rect in view space (origin top-left) into a normalized rect,
    /// given the frame the page occupies in that same view space. Clamped to 0..1.
    /// </summary>
    public PdfRect Normalize(PdfRect viewRect, PdfRect pageFrame)
    {
        if (pageFrame.Width <= 0 || pageFrame.Height <= 0) return PdfRect.Zero;
        var nx = (viewRect.MinX - pageFrame.MinX) / pageFrame.Width;
        var ny = (viewRect.MinY - pageFrame.MinY) / pageFrame.Height;
        var nw = viewRect.Width / pageFrame.Width;
        var nh = viewRect.Height / pageFrame.Height;
        return Clamp01(new PdfRect(nx, ny, nw, nh));
    }

    /// <summary>Inverse of <see cref="Normalize"/> — place a normalized rect back into a view frame.</summary>
    public PdfRect ViewRect(PdfRect n, PdfRect pageFrame) => new(
        pageFrame.MinX + n.MinX * pageFrame.Width,
        pageFrame.MinY + n.MinY * pageFrame.Height,
        n.Width * pageFrame.Width,
        n.Height * pageFrame.Height);

    // ── Normalized ↔ PDF user space ──────────────────────────────────────────

    /// <summary>
    /// Map a normalized displayed rect (origin top-left) to a PDF user-space rect
    /// (origin bottom-left, unrotated, cropBox-offset) for /Rect.
    /// </summary>
    public PdfRect UserSpaceRect(PdfRect n)
    {
        var d = DisplayedSize;
        // Normalized (top-left) → displayed points (bottom-left origin).
        var dx = n.MinX * d.Width;
        var dw = n.Width * d.Width;
        var dh = n.Height * d.Height;
        var dy = d.Height - (n.MinY * d.Height) - dh; // flip y
        // The two opposite corners of the box in displayed space.
        var c0 = DisplayedToLocal(new PdfPoint(dx, dy));
        var c1 = DisplayedToLocal(new PdfPoint(dx + dw, dy + dh));
        var minX = Math.Min(c0.X, c1.X);
        var minY = Math.Min(c0.Y, c1.Y);
        var maxX = Math.Max(c0.X, c1.X);
        var maxY = Math.Max(c0.Y, c1.Y);
        return new PdfRect(
            CropBox.MinX + minX,
            CropBox.MinY + minY,
            maxX - minX,
            maxY - minY);
    }

    /// <summary>Inverse: PDF user-space rect → normalized displayed rect (top-left).</summary>
    public PdfRect NormalizedRect(PdfRect r)
    {
        var local0 = new PdfPoint(r.MinX - CropBox.MinX, r.MinY - CropBox.MinY);
        var local1 = new PdfPoint(r.MaxX - CropBox.MinX, r.MaxY - CropBox.MinY);
        var d0 = LocalToDisplayed(local0);
        var d1 = LocalToDisplayed(local1);
        var dminX = Math.Min(d0.X, d1.X);
        var dmaxX = Math.Max(d0.X, d1.X);
        var dminY = Math.Min(d0.Y, d1.Y);
        var dmaxY = Math.Max(d0.Y, d1.Y);
        var d = DisplayedSize;
        if (d.Width <= 0 || d.Height <= 0) return PdfRect.Zero;
        var nx = dminX / d.Width;
        var nw = (dmaxX - dminX) / d.Width;
        var nh = (dmaxY - dminY) / d.Height;
        // Flip y back to top-left origin.
        var ny = 1.0 - (dmaxY / d.Height);
        return new PdfRect(nx, ny, nw, nh);
    }

    // ── Rotation core (page-local, bottom-left origin) ───────────────────────

    /// <summary>Displayed-space point (bottom-left) → page-local point (unrotated, bottom-left).</summary>
    private PdfPoint DisplayedToLocal(PdfPoint p)
    {
        var w = CropBox.Width;
        var h = CropBox.Height;
        return Rotation switch
        {
            90 => new PdfPoint(w - p.Y, p.X),
            180 => new PdfPoint(w - p.X, h - p.Y),
            270 => new PdfPoint(p.Y, h - p.X),
            _ => p,
        };
    }

    /// <summary>Page-local point → displayed-space point (inverse of <see cref="DisplayedToLocal"/>).</summary>
    private PdfPoint LocalToDisplayed(PdfPoint p)
    {
        var w = CropBox.Width;
        var h = CropBox.Height;
        return Rotation switch
        {
            90 => new PdfPoint(p.Y, w - p.X),
            180 => new PdfPoint(w - p.X, h - p.Y),
            270 => new PdfPoint(h - p.Y, p.X),
            _ => p,
        };
    }

    private static PdfRect Clamp01(PdfRect r)
    {
        var x = Math.Max(0, Math.Min(1, r.MinX));
        var y = Math.Max(0, Math.Min(1, r.MinY));
        var w = Math.Max(0, Math.Min(1 - x, r.Width));
        var h = Math.Max(0, Math.Min(1 - y, r.Height));
        return new PdfRect(x, y, w, h);
    }

    public bool Equals(CoordinateMapper other) => CropBox.Equals(other.CropBox) && Rotation == other.Rotation;
    public override bool Equals(object? obj) => obj is CoordinateMapper m && Equals(m);
    public override int GetHashCode() => HashCode.Combine(CropBox, Rotation);
}
